### PR TITLE
warp: Use more meaningful types instead of `Bool`

### PR DESCRIPTION
**Pull request by sol**
The overuse of booleans for all kinds of things makes it somewhat hard to follow the code that implements HTTP/1.1 keep-alive.  This is a step towards trying to improve the situation.

### CODE REVIEW

File: warp/Network/Wai/Handler/Warp/Request.hs [hunk 1]
```diff
@@ -50,11 +51,13 @@ import Network.Wai.Handler.Warp.Settings (
 
 ----------------------------------------------------------------
 
+-- | first request on this connection?
+data FirstRequest = FirstRequest | SubsequentRequest
+
 -- | Receiving a HTTP request from 'Connection' and parsing its header
 --   to create 'Request'.
 recvRequest
```
Comment (sol) [inline warp/Network/Wai/Handler/Warp/Request.hs:59]: Apparently, this  is re-exported from `Network.Wai.Handler.Warp.Internal`.  @kazu-yamamoto @Vlix what are the stability guarantees for that module?

Comment (kazu-yamamoto) [reply warp/Network/Wai/Handler/Warp/Request.hs:59]: We can change `Internal` at anytime without major version up.

**Comment by kazu-yamamoto** [comment]
I have fixed CI and rebase this PR onto `master`.
CI failed again.
I fixed Bench but still failing.

https://github.com/kazu-yamamoto/wai/actions/runs/10244402661/job/28337330217

**Comment by Vlix** [comment]
I'll take a look at it tomorrow 👍

**Comment by kazu-yamamoto** [comment]
With my refreshed head, it is easy to fix this.
I have merged this PR after rebase with my fix.
Thank you for your contribution!
